### PR TITLE
DNN: Fix Squeeze to remove all size-1 dims when axes is empty

### LINKED ISSUE #28424: [DNN/ONNX] Squeeze operator fails when axes is empty
### System Information

- **OpenCV Version**: 4.13.0
- **OS**: Linux
- **Python**: 3.10
- **ONNX Opset**: 11
- **Comparison**: ONNX Runtime (works correctly)

### Detailed description

OpenCV's DNN module incorrectly handles the ONNX `Squeeze` operator when all axes are squeezed (i.e., removing all dimensions of size 1). While ONNX Runtime correctly produces a `[64, 64]` output tensor, OpenCV DNN retains the original `[1, 1, 64, 64]` shape.

### Expected Behavior
When a `Squeeze` node removes all size-1 dimensions:
- **Input shape**: `[1, 1, 64, 64]`
- **Expected output**: `[64, 64]`

### Actual Behavior (OpenCV DNN)
- **Actual output**: `[1, 1, 64, 64]` (no dimensions removed)

### ONNX Runtime Behavior (Reference)
- **Output**: `[64, 64]` ✓ (correct)

## Steps to Reproduce

```python
import torch
import torch.nn as nn
import cv2
import onnxruntime as ort

class SqueezeAllDims(nn.Module):
    def forward(self, x):
        return torch.squeeze(x)  # Remove all size-1 dims

model = SqueezeAllDims()
model.eval()

input_tensor = torch.randn(1, 1, 64, 64)

# Export to ONNX
torch.onnx.export(model, input_tensor, "squeeze_bug.onnx",
                 input_names=['input'], output_names=['output'],
                 opset_version=11)

# ONNX Runtime (correct)
ort_session = ort.InferenceSession("squeeze_bug.onnx")
ort_out = ort_session.run(None, {'input': input_tensor.numpy()})[0]
print(f"ONNX Runtime shape: {ort_out.shape}")  # (64, 64) ✓

# OpenCV DNN (bug)
net = cv2.dnn.readNetFromONNX("squeeze_bug.onnx")
net.setInput(input_tensor.numpy())
cv_out = net.forward()
print(f"OpenCV DNN shape: {cv_out.shape}")  # (1, 1, 64, 64) ✗
```

## Analysis

The issue likely stems from the `Squeeze` layer implementation in OpenCV DNN:
- **File**: `modules/dnn/src/layers/reshape_layer.cpp` (SqueezeLayer)
- **Root cause**: When the `axes` attribute is empty or not specified, the layer should squeeze **all** dimensions of size 1, but OpenCV appears to skip this operation

### ONNX Specification
According to [ONNX Squeeze operator spec](https://onnx.ai/onnx/operators/onnx__Squeeze.html):
> If `axes` is not provided, all the single dimensions will be removed from the shape.

OpenCV DNN does not correctly implement this fallback behavior.

## Proposed Fix

In `modules/dnn/src/layers/reshape_layer.cpp`, the `Squeeze` layer should:
1. Check if `axes` parameter is empty
2. If empty, automatically detect and remove all dimensions where `size == 1`
3. Update output shape accordingly

Example pseudocode:
```cpp
if (axes.empty()) {
    // Squeeze ALL dimensions of size 1
    for (int i = 0; i < input_shape.size(); i++) {
        if (input_shape[i] != 1) {
            output_shape.push_back(input_shape[i]);
        }
    }
} else {
    // Squeeze only specified axes
    // ... existing logic ...
}
```
## Attachments
- ONNX file: `squeeze_bug.onnx.zip`
[squeeze_bug.onnx.zip](https://github.com/user-attachments/files/24695873/squeeze_bug.onnx.zip)

## Impact
This affects any model using `torch.squeeze()` without axis specification, including:
- Models with dynamic batch sizes normalized to 1
- Auto-encoder bottleneck layers
- Channel-wise global pooling outputs

**Note**: I'm ready to submit a PR if this issue is confirmed. Please advise on the best approach.

### Issue submission checklist

- [x] I report the issue, it's not a question
- [x] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [x] I updated to the latest OpenCV version and the issue is still there
- [x] There is reproducer code and related data files (videos, images, onnx, etc)

### PR DESCRIPTION
**Pull request by Anemptyship**
Fixes #28424
OpenCV Extra: [opencv/opencv_extra#1308](https://github.com/opencv/opencv_extra/pull/1308)

This PR fixes the ONNX Squeeze operator to correctly remove all size-1 dimensions when `axes` is not provided, conforming to the ONNX specification.

### Details
Per [ONNX Squeeze specification](https://onnx.ai/onnx/operators/onnx__Squeeze.html):
> 'If axes is not provided, all the single dimensions will be removed from the shape.'

Previously, OpenCV DNN would not remove any dimensions in this case, causing shape mismatch errors with models like LaMa (inpainting).

### Example
```python
# Input: [1, 1, 2, 4]
# Squeeze with no axes attribute

# Before: [1, 1, 2, 4] ✗ (No change)
# After:  [2, 4] ✓ (matches ONNX Runtime)
```

### Tests
Added `testONNXModels("squeeze_no_axes")` which validates this behavior with new test data.
opencv_extra_pr=opencv/opencv_extra#1324

### Pull Request Readiness Checklist
See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch (4.x for bug fixes)
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
  - Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by asmorkalov** [comment]
@Anemptyship Thanks for the patch. Could you create a small C++ test for it with ONNX file like described in the issue. You need yo export the model with test input and output and create a PR with the same branch name as here to opencv_extra repo. Also please place the code to generate the model to https://github.com/opencv/opencv_extra/blob/4.x/testdata/dnn/onnx/generate_onnx_models.py. Please use https://github.com/opencv/opencv_extra/pull/1299 as example. The PR is done for TFLite model. You should get similar, but for ONNX.

**Comment by Anemptyship** [comment]
> @Anemptyship Thanks for the patch. Could you create a small C++ test for it with ONNX file like described in the issue. You need yo export the model with test input and output and create a PR with the same branch name as here to opencv_extra repo. Also please place the code to generate the model to https://github.com/opencv/opencv_extra/blob/4.x/testdata/dnn/onnx/generate_onnx_models.py. Please use [opencv/opencv_extra#1299](https://github.com/opencv/opencv_extra/pull/1299) as example. The PR is done for TFLite model. You should get similar, but for ONNX.

Thanks for the review!

I have added the requested test case.
- **Test case**: `squeeze_no_axes` (added to [test_onnx_importer.cpp](cci:7://file:///home/ben/dev/opencv-fork/modules/dnn/test/test_onnx_importer.cpp:0:0-0:0))
- **Test data**: Added generation script and models to `opencv_extra`
  - Input: `[1, 1, 2, 4]`
  - Expected Output: `[2, 4]` (All size-1 dimensions are removed)
OpenCV Extra: [opencv/opencv_extra#1308](https://github.com/opencv/opencv_extra/pull/1308)

**Comment by asmorkalov** [comment]
@abhishek-gola Could you take a look on the solution?

**Comment by asmorkalov** [comment]
@abhishek-gola Do we need similar patch for 5.x?

File: modules/dnn/test/test_onnx_importer.cpp [hunk 1]
```diff
@@ -1187,6 +1187,7 @@ TEST_P(Test_ONNX_layers, Squeeze)
         applyTestTag(CV_TEST_TAG_DNN_SKIP_IE_MYRIAD, CV_TEST_TAG_DNN_SKIP_IE_NN_BUILDER);
     testONNXModels("squeeze");
     testONNXModels("squeeze_axes_op13");
+    testONNXModels("squeeze_no_axes");
```
Comment (asmorkalov) [inline modules/dnn/test/test_onnx_importer.cpp:1192]: The test does not check expected output tensor shape.

**Comment by abhishek-gola** [comment]
> @abhishek-gola Do we need similar patch for 5.x?

@asmorkalov, We don’t need this patch in the 5.x branch, since the behavior is already handled by a dedicated layer implementation in squeeze_layer.cpp:
https://github.com/opencv/opencv/blob/5.x/modules/dnn/src/layers/squeeze_layer.cpp#L55-L58

**Comment by abhishek-gola** [comment]
@Anemptyship, 
```
Run cd $HOME/build && python3 $HOME/scripts/warnings-handling.py
/home/ci/opencv/modules/dnn/test/test_onnx_importer.cpp:107:12: warning: suggest explicit braces to avoid ambiguous 'else' [-Wdangling-else]

Build has 1 warnings.
```

Fix these warnings.

**Comment by asmorkalov** [comment]
Build warning:
```
Warning check
  Warning: /home/ubuntu/opencv/modules/dnn/test/test_onnx_importer.cpp:107:12: warning: suggest explicit braces to avoid ambiguous 'else' [-Wdangling-else]
```

**Comment by asmorkalov** [comment]
```
Warning: /home/ubuntu/opencv/modules/dnn/test/test_onnx_importer.cpp:107:12: warning: suggest explicit braces to avoid ambiguous 'else' [-Wdangling-else]
```

**Comment by asmorkalov** [comment]
@dkurt Could you take a look on the patch?

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 2]
```diff
@@ -2163,6 +2129,14 @@ void ONNXImporter::parseSqueeze(LayerParams& layerParams, const opencv_onnx::Nod
...
             CV_Error(Error::StsNotImplemented, cv::format("ONNX/Squeeze: doesn't support non-constant 'axes' input"));
     }
+    else
+    {
+        for (int i = 0; i < inpShape.size(); ++i)
+        {
+            if (inpShape[i] == 1)
+                maskedAxes[i] = true;
+        }
+    }
```
Comment (asmorkalov) [inline modules/dnn/src/onnx/onnx_importer.cpp:2139]: DNN should squeeze only leading ones. In case of shape (64, 1, 32) the internal `1` should be preserved. It's not true for now.

Comment (Anemptyship) [reply modules/dnn/src/onnx/onnx_importer.cpp:2139]: @asmorkalov Thanks for the feedback. Just to clarify — the [ONNX Squeeze spec](https://onnx.ai/onnx/operators/onnx__Squeeze.html) states that when axes is not provided, all size-1 dimensions should be removed. Should I follow that, or do you prefer the "leading only" approach for a specific reason?

**Comment by asmorkalov** [comment]
@vpisarev @abhishek-gola @dkurt Could you comment?

File: modules/dnn/src/onnx/onnx_importer.cpp (hunk 2 above)
Comment (dkurt) [reply modules/dnn/src/onnx/onnx_importer.cpp:2139]: @Anemptyship the added test demonstrates only leading 1 removal. Please also add a test case with internal 1 dimensions removed by onnxruntime. For example, `[1, 2, 1, 4] -> [2, 4]`

File: modules/dnn/test/test_onnx_importer.cpp [hunk 3]
```diff
@@ -103,6 +104,11 @@ class Test_ONNX_layers : public DNNTestLayer
             net.setInput(inps[i], inputNames[i]);
         Mat out = net.forward("");
 
+        if (!expectedOutputShape.empty())
+        {
+            ASSERT_EQ(shape(out), expectedOutputShape);
```
Comment (dkurt) [inline modules/dnn/test/test_onnx_importer.cpp]: Why not compare with `shape(ref)`?

Comment (Anemptyship) [reply modules/dnn/test/test_onnx_importer.cpp]: Thanks for the review! Done. Changed to use `shape(ref)`.

File: modules/dnn/src/onnx/onnx_importer.cpp (hunk 2 above)
Comment (Anemptyship) [reply modules/dnn/src/onnx/onnx_importer.cpp:2139]: @dkurt Thanks! Done. Updated the test input to [1, 2, 1, 4] to cover internal size-1 dims as well.

Updated test data in opencv_extra: https://github.com/Anemptyship/opencv_extra/commit/745bcfa
